src/flow/src/transform/plan.rs

```diff
@@ -79,8 +79,23 @@ impl TypedPlan {
                 };
 
                 let mut exprs: Vec<TypedExpr> = vec![];
+                // because this `input.schema` is incorrect for pre-expand substrait plan, so we have to use schema before expand multi-value
+                // function to correctly transform it, and late rewrite it
+                let schema_before_expand = {
+                    let input_schema = input.schema.clone();
+                    let auto_columns: HashSet<usize> =
+                        HashSet::from_iter(input_schema.typ().auto_columns.clone());
+                    let not_auto_added_columns = (0..input_schema.len()?)
+                        .filter(|i| !auto_columns.contains(i))
+                        .collect_vec();
+                    let mfp = MapFilterProject::new(input_schema.len()?)
+                        .project(not_auto_added_columns)?
+                        .into_safe();
+
+                    input_schema.apply_mfp(&mfp)?
+                };
```

Comment:
The `from_substrait_rel` method is quite large. Can we split its logic into multiple methods? This might make it clearer, for example(Maybe next PR):
```Rust
pub fn from_substrait_rel(
    ctx: &mut FlownodeContext,
    rel: &Rel,
    extensions: &FunctionExtensions,
) -> Result<TypedPlan, Error> {
    match &rel.rel_type {
        Some(RelType::Project(p)) => handle_project(ctx, p, extensions),
        Some(RelType::Filter(filter)) => handle_filter(ctx, filter, extensions),
        Some(RelType::Read(read)) => handle_read(ctx, read),
        Some(RelType::Aggregate(agg)) => TypedPlan::from_substrait_agg_rel(ctx, agg, extensions),
        _ => not_impl_err!("Unsupported relation type: {:?}", rel.rel_type),
    }
}
```